### PR TITLE
Remove assert in favor of error message

### LINKED ISSUE #29795: Don't use `assert` in Spack
### Steps to reproduce

The file [`util/url.py`](https://github.com/spack/spack/blob/58a32b04d90da7d2edc4388e78feb84abe8fb089/lib/spack/spack/util/url.py#L347) contains an `assert` statement. If it triggers, then no meaningful error message is generated. Instead, Spack just aborts.

This is triggered when using a URL with an unknown prefix. I encountered this when debugging why CVS repositories don't work any more (see https://github.com/spack/spack/pull/29793).

### Error message

None, see above, Spack just aborts.

### Information on your system

```
$ spack debug report
* **Spack:** 0.18.0.dev0 (58a32b04d90da7d2edc4388e78feb84abe8fb089)
* **Python:** 3.8.9
* **Platform:** darwin-monterey-skylake
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by johnwparent**
Switches the less than helpful assert statement into a more verbose and debug-able error message.

Fixes: #29795

### CODE REVIEW

**Review by alalazo — APPROVED** [review]

**Review by alalazo — CHANGES_REQUESTED** [review]
Can you check the failure in CI? It seems `spack create` triggers this new error.

**Review by alalazo — DISMISSED** [review]

File: lib/spack/spack/test/cmd/create.py [hunk 1]
```diff
@@ -16,6 +16,15 @@
 create = SpackCommand('create')
 
 
+def _validate_url(url):
+    return
+
+
+@pytest.fixture(autouse=True)
+def url_check(monkeypatch):
+    monkeypatch.setattr(spack.util.url, 'require_url_format', _validate_url)
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/create.py]: The rebase turned off the auto merge, I'd merge it myself but I'm personally curious whether these tests can be made to pass without this monkeypatching. IIRC we spoke about a few cases like this in https://github.com/spack/spack/pull/27021 and my general goal was to not use it where possible. I notice that @alalazo has already approved this, so feel free to merge it if this is not a concern

File: lib/spack/spack/util/url.py [hunk 2]
```diff
@@ -344,7 +344,8 @@ def parse_git_url(url):
 
 def require_url_format(url):
     ut = re.search(r'^(file://|http://|https://|ftp://|s3://|gs://|ssh://|git://|/)', url)
-    assert ut is not None
+    if not ut:
+        raise ValueError('Invalid url format from url: %s' % url)
```
Comment (scheibelp) [inline lib/spack/spack/util/url.py:348]: I think on account of this, we can also modify `cmd.create.get_versions` to no longer catch `AssertionError` (which IMO should be done in this PR) - does that sound reasonable to you?

Comment (johnwparent) [reply lib/spack/spack/util/url.py:348]: Yes it does, will make those changes.

**Review by scheibelp — APPROVED** [review]

**Pull request merged**